### ISSUE TITLE
Dask support

### ISSUE THREAD

**Issue by aplavin** [tqdm/tqdm#278]
Dask itself has a basic progressbar, but tqdm is certainly better - so I made a basic wrapper:

``` python
from dask.callbacks import Callback

class ProgressBar(Callback):
    def _start_state(self, dsk, state):
        self._tqdm = tqdm_notebook(total=sum(len(state[k]) for k in ['ready', 'waiting', 'running', 'finished']))

    def _posttask(self, key, result, dsk, state, worker_id):
        self._tqdm.update(1)

    def _finish(self, dsk, state, errored):
        pass
```

Usage (the same as dask progressbar):

``` python
with ProgressBar():
    ... .compute()
```

Does it belong to `tqdm`, what to you think? Also, any further suggestions/improvements?

**Comment by lrq3000** [comment]
Thank's for the contribution, that's great!

I did not use dask myself yet, so I cannot provide a precise feedback.

Just a question: in dask, are each element processed with equal time, or
can they take different amount of time? Also, are elements full objects
(like a full dataframe, so the callback will happen after the dataframe
gets entirely processed), or are they exploded into subelements (eg,
dataframe being exploded into rows, and the rows are the elements composing
a dask object)?

2016-10-01 13:30 GMT+02:00 Alexander <notifications@github.com>:

**Comment by aplavin** [comment]
@lrq3000 well, in dask "tasks" are chunks - the count of which you usually specify when creating dask  dataframe or something. Each chunk in case of dask dataframe is a pandas dataframe, and we can have callback only once for each chunk. I don't think it's possible to get the count of individual rows or items in the general case because on the low level dask is just a scheduler for various multiprocessing stuff.

**Comment by lrq3000** [comment]
Ok Alex, then I don't think it's possible to do better for now, you already
made a great wrapper :) We will add it to the readme! (or maybe this can be
monkeypatched like we did for pandas?).

2016-10-01 14:40 GMT+02:00 Alexander <notifications@github.com>:

**Comment by aplavin** [comment]
@lrq3000 I don't think monkeypatching makes sense here - the very basic progressbar built-in in dask is actually used the same way as I wrote above:
```
from dask.diagnostics import ProgressBar

with ProgressBar():
    ... .compute()
```
So, something like `from tqdm.dask import ProgressBar` would be ok, I think.

**Comment by lrq3000** [comment]
Ok thank you Alexander, indeed it's better to follow the dask's API, we
will do that :)

2016-10-01 16:37 GMT+02:00 Alexander <notifications@github.com>:

**Comment by aplavin** [comment]
So I guess I'll make a PR soon.

**Comment by lrq3000** [comment]
If you want to be accredited as author in the commits then yes, feel free
to make a PR, we will gladly merge it :) But note that it should be a
submodule, not in the core tqdm!

**Comment by norandom** [comment]
This is very interesting. Has this been merged into a current release?

**Comment by lrq3000** [comment]
Not yet merged but it will be.
Le 16 Oct. 2016 11:10, "Marius Ciepluch" <notifications@github.com> a
écrit :

**Comment by lrq3000** [comment]
To be more precise: this will get merged after #198, and before that I'm waiting for someone to review [some bugfixes and enhancements](https://github.com/tqdm/tqdm/issues?utf8=%E2%9C%93&q=is%3Aopen%20label%3Ato-review%20%20-label%3Asubmodule). So don't hold your breath, it will be merged eventually before Christmas I think, but I'm not sure when.

**Comment by alexifm** [comment]
Hey, wanted to follow up on this and show a dask distributed version for tqdm.  This is based off the progress bars built here: https://github.com/dask/distributed/blob/master/distributed/diagnostics/progressbar.py 
```
import tqdm
from tornado.ioloop import IOLoop
from distributed.utils import LoopRunner, is_kernel
from distributed.client import futures_of
from distributed.diagnostics.progressbar import ProgressBar

class TqdmProgressBar(ProgressBar):
    def __init__(
        self,
        keys,
        scheduler=None,
        interval="100ms",
        loop=None,
        complete=True,
        start=True,
        **tqdm_kwargs
    ):
        super(TqdmProgressBar, self).__init__(
            keys, scheduler, interval, complete)
        self.tqdm = tqdm.tqdm(keys, **tqdm_kwargs)
        self.loop = loop or IOLoop()

        if start:
            loop_runner = LoopRunner(self.loop)
            loop_runner.run_sync(self.listen)

    def _draw_bar(self, remaining, all, **kwargs):
        update_ct = (all - remaining) - self.tqdm.n
        self.tqdm.update(update_ct)

    def _draw_stop(self, **kwargs):
        pass

class TqdmNotebookProgress(ProgressBar):
    def __init__(
        self,
        keys,
        scheduler=None,
        interval="100ms",
        loop=None,
        complete=True,
        start=True,
        **tqdm_kwargs
    ):
        super(TqdmProgressWidget, self).__init__(
            keys, scheduler, interval, complete)
        self.tqdm = tqdm.tqdm_notebook(keys, **tqdm_kwargs)
        self.loop = loop or IOLoop()

        if start:
            loop_runner = LoopRunner(self.loop)
            loop_runner.run_sync(self.listen)

    def _draw_bar(self, remaining, all, **kwargs):
        update_ct = (all - remaining) - self.tqdm.n
        self.tqdm.update(update_ct)

    def _draw_stop(self, **kwargs):
        pass

def tqdm_dask(futures, **kwargs):
    notebook = is_kernel()
    futures = futures_of(futures)
    if not isinstance(futures, (set, list)):
        futures = [futures]
    if notebook:
        return TqdmNotebookProgress(futures, **kwargs)
    else:
        TqdmProgressBar(futures, **kwargs)
```
If you run in a notebook, you'll get the widget version.
```
from dask.distributed import Client, LocalCluster

def square(x):
    return x ** 2

def test_progress(n):
    A = client.map(square, range(n))
    TqdmProgressBar(A)
    results = client.gather(A)
    return results

def test_progress_notebook(n):
    A = client.map(square, range(n))
    TqdmNotebookProgress(A)
    results = client.gather(A)
    return results

def test_tqdm_dask(n):
    A = client.map(square, range(n))
    tqdm_dask(A)
    results = client.gather(A)
    return results

cluster = LocalCluster()
client = Client(cluster)

results = test_progress(10000)
results = test_progress_notebook(10000)
results = test_tqdm_dask(10000)
```

**Comment by thomasaarholt** [comment]
Since it wasn't mentioned in this issue, I'd just like to add to the original post that one can now use `tqdm.auto.tqdm` instead of `tqdm.tqdm_notebook`, so the code from OP becomes:
```python
from dask.callbacks import Callback
from tqdm.auto import tqdm

class ProgressBar(Callback):
    def _start_state(self, dsk, state):
        self._tqdm = tqdm(total=sum(len(state[k]) for k in ['ready', 'waiting', 'running', 'finished']))

    def _posttask(self, key, result, dsk, state, worker_id):
        self._tqdm.update(1)

    def _finish(self, dsk, state, errored):
        pass
```
Usage (the same as dask progressbar):

```python
with ProgressBar():
    ... .compute()
```
which results in the lovely widget-based progressbar in the notebook:
<img src="https://user-images.githubusercontent.com/2721423/85793594-88883900-b735-11ea-8c72-b9705d5263b0.png" width="500">
 and the standard one in command line:
<img src="https://user-images.githubusercontent.com/2721423/85793899-05b3ae00-b736-11ea-9ecc-0231a9c221cd.png" width="1000">

**Comment by DevEliran** [comment]
@thomasaarholt Is there a way to support a custom description for this ProgressBar callback?

**Comment by thomasaarholt** [comment]
Sure! Just modify the `tqdm(...)` command to be `tqdm(..., `desc='your description'`)`. See the [docs ](https://github.com/tqdm/tqdm#usage)for *lots* more options.

**Comment by DevEliran** [comment]
@thomasaarholt I meant like how you would pass as an argument a description to 
`with ProgressBar():`  
I didn't mean to add some fixed text as a description like that  
```
def _start_state(self, dsk, state):
        self._tqdm = tqdm(total=sum(len(state[k]) for k in ['ready', 'waiting', 'running', 'finished']), desc="some description")
```

I would want that every context manager would have it's own description; how can I modify ProgressBar class to allow such functionallity?

For example

```
with ProgressBar(desc="doing some stuff"):
    ----do some stuff---

with ProgressBar(desc="doing other stuff"):
    ----do other stuff----
```

**Comment by thomasaarholt** [comment]
Ah, sure! I'm not able to check right now, but I reckon that if you add a

```python
__init__(self, desc):
    self.desc=desc
```
And inside tqdm() set `desc=self.desc`, your examples should work. Let me know if it does.

**Comment by casperdcl** [comment]
try out #1079 (`tqdm.dask`) - this uses the same approach as `tqdm.keras`, including the recent fixes.

**Comment by deschman** [comment]
I have been working with #1079 and have found it is not compatible with the most recent dask release 2021.2.0.

**Comment by thomasaarholt** [comment]
I came back to this after improving a StackOverflow post asking the same. The following is a working solution for dask progressbar based on tqdm:

```python
from dask.callbacks import Callback
from tqdm.auto import tqdm

class ProgressBar(Callback):
    def __init__(self, desc=""):
        self.desc = desc

    def _start_state(self, dsk, state):
        self._tqdm = tqdm(total=sum(len(state[k]) for k in ['ready', 'waiting', 'running', 'finished']), desc=self.desc)

    def _posttask(self, key, result, dsk, state, worker_id):
        self._tqdm.update(1)

    def _finish(self, dsk, state, errored):
        pass
```

Use it as:
```python
with ProgressBar("your description"):
    arr.compute() # your Dask computation here
```

**Comment by casperdcl** [comment]
er https://github.com/tqdm/tqdm/pull/1079 (vis. https://github.com/tqdm/tqdm#dask-integration & https://tqdm.github.io/docs/dask/) is probably better

**Comment by thomasaarholt** [comment]
Ah! Brilliant, I will use that instead!

**Comment by dendihandian** [comment]
how apply `ProgressBar` on `apply()` like `result_df = raw_df.apply(lambda row: process_row(row), axis=1).compute()`?
